File: samples/java/Bingo/Server/Matchmaking/src/main/java/systems/zlink/samples/bingo/server/matchmaking/ReserveBingoRoomHandler.java
package systems.zlink.samples.bingo.server.matchmaking;

import systems.zlink.framework.spots.ZLinkSpotRequestHandler;
import systems.zlink.samples.bingo.shared.contracts.Messages;

import java.util.concurrent.CompletionStage;

public final class ReserveBingoRoomHandler
        implements ZLinkSpotRequestHandler<
                BingoMatchmaker, Messages.ReserveBingoRoomReq, Messages.ReserveBingoRoomRes> {
    private final RedisBingoMatchReservationStore reservations;

    public ReserveBingoRoomHandler(RedisBingoMatchReservationStore reservations) {
        this.reservations = reservations;
    }

    @Override
    public CompletionStage<Messages.ReserveBingoRoomRes> handle(
            BingoMatchmaker spot, Messages.ReserveBingoRoomReq request) {
        spot.beginRequest();
        return reservations.reserve(request).whenComplete((ignored, failure) -> spot.endRequest());
    }
}

File: samples/kotlin/Bingo/Server/Matchmaking/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/matchmaking/ReserveBingoRoomHandler.kt
package systems.zlink.samples.kotlin.bingo.server.matchmaking

import java.util.concurrent.CompletionStage
import systems.zlink.framework.spots.ZLinkSpotRequestHandler
import systems.zlink.samples.kotlin.bingo.shared.contracts.ReserveBingoRoomReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.ReserveBingoRoomRes

class ReserveBingoRoomHandler(private val reservations: RedisBingoMatchReservationStore) :
    ZLinkSpotRequestHandler<BingoMatchmaker, ReserveBingoRoomReq, ReserveBingoRoomRes> {
    override fun handle(
        spot: BingoMatchmaker,
        request: ReserveBingoRoomReq,
    ): CompletionStage<ReserveBingoRoomRes> {
        spot.beginRequest()
        return reservations.reserve(request).whenComplete { _, _ -> spot.endRequest() }
    }
}
